internal/objectstore/remote/retriever.go

```diff
@@ -0,0 +1,298 @@
+			if err := res.err; errors.Is(err, BlobNotFound) {
+				if len(results) == len(remotes) {
+					return nil, -1, BlobNotFound
+				}
+				continue
+			} else if err != nil {
+				// If there is an error that is not BlobNotFound, return it
+				return nil, -1, err
+			}
+
+			// Stop all the other retrievers, we don't want to cancel the
+			// retriever that is currently being used, as that will cause the
+			// reader to be closed.
+			for _, retriever := range retrievers {
+				if retriever.index == res.index {
+					continue
+				}
+
+				retriever.Kill()
+
+				// Don't wait for them to stop, when closing the reader, we will
+				// wait for them to stop then. We just want them to stop
+				// processing the request as soon as possible.
+			}
+
+			return &retrieverReaderCloser{
+				reader: res.reader,
+				closer: func() {
+					r.stopAllRetrievers(ctx, retrievers)
+				},
+			}, res.size, nil
+		}
+	}
+}
+
+// Kill stops the BlobRetriever.
+func (r *BlobRetriever) Kill() {
+	r.tomb.Kill(nil)
+}
+
+// Wait waits for the BlobRetriever to stop.
+func (r *BlobRetriever) Wait() error {
+	return r.tomb.Wait()
+}
+
+func (r *BlobRetriever) loop() error {
+	select {
+	case <-r.tomb.Dying():
+		return tomb.ErrDying
+	}
+}
+
+// stopAllRetrievers stops all the retrievers and waits for them to stop. This
+// ensures that there are no dangling goroutines.
+func (r *BlobRetriever) stopAllRetrievers(ctx context.Context, retrievers []*retriever) {
+	// Kill 'Em All.
+	for _, retriever := range retrievers {
+		retriever.Kill()
+		// Wait for them to stop, we don't want to leave any hanging.
+		if err := retriever.Wait(); err != nil {
+			r.logger.Errorf(ctx, "failed to stop blob retriever: %v", err)
+		}
+	}
+}
+
+type retriever struct {
+	tomb tomb.Tomb
+
+	index           int
+	remote          apiremotecaller.RemoteConnection
+	newObjectClient NewObjectClientFunc
+	clock           clock.Clock
+	logger          logger.Logger
+}
+
+func newRetriever(index int, remote apiremotecaller.RemoteConnection, newObjectClient NewObjectClientFunc, clock clock.Clock, logger logger.Logger) *retriever {
+	t := &retriever{
+		index:           index,
+		remote:          remote,
+		newObjectClient: newObjectClient,
+		clock:           clock,
+		logger:          logger,
+	}
+
+	t.tomb.Go(func() error {
+		<-t.tomb.Dying()
+		return tomb.ErrDying
+	})
+
+	return t
+}
+
+// Retrieve requests a blob from the remote API server, if there is not a remote
+// connection, it will retry a number of times before returning an error. This
+// is to allow time for the API connection to come up on a new HA node, or after
+// a restart. If the blob isn't found or any other non-retryable error it will
+// return an error right away. If the context is cancelled, it will stop
+// processing the request as soon as possible.
+func (t *retriever) Retrieve(ctx context.Context, namespace, sha256 string) (io.ReadCloser, int64, error) {
+	ctx, cancel := context.WithCancel(ctx)
```

Comment:
If `t` needs a tomb, should we not use it here?

```suggestion
	ctx, cancel := context.WithCancel(t.tomb.Context(ctx))
```